File: modules/count.c
#include <stdlib.h>
#include "soundpipe.h"

int sp_count_create(sp_count **p)
{
    *p = malloc(sizeof(sp_count));
    return SP_OK;
}

int sp_count_destroy(sp_count **p)
{
    free(*p);
    return SP_OK;
}

int sp_count_init(sp_data *sp, sp_count *p)
{
    p->count = 4;
    p->curcount = -1;
    return SP_OK;
}

int sp_count_compute(sp_data *sp, sp_count *p, SPFLOAT *in, SPFLOAT *out)
{
    int i;
    if(*in){
        p->curcount = (p->curcount + 1) % p->count;
    }
    *out = p->curcount;
    return SP_OK;
}

File: modules/data/count.lua
sptbl["count"] = {

    files = {
        module = "count.c",
        header = "count.h",
        example = "ex_count.c",
    },

    func = {
        create = "sp_count_create",
        destroy = "sp_count_destroy",
        init = "sp_count_init",
        compute = "sp_count_compute",
    },

    params = {
        optional = {
            {
                name = "count",
                type = "SPFLOAT",
                description = "Number to count up to (count - 1). Decimal points will be truncated.",
                default = 4
            },
        },
    },

    modtype = "module",

    description = [[Trigger-based fixed counter. The signal output will count from 0 to [N-1], and then
repeat itself. ]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "trig",
            description = "When non-zero, will increment. When the output value is N - 1, it will loop back to zero."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Signal output."
        },
    }

}
